tests/utilities/test_fetching.py

```diff
@@ -572,3 +574,54 @@ def training_step(self, dataloader_iter):
     durations = profiler.recorded_durations[key]
     assert len(durations) == 2  # 2 polls in training_step
     assert all(d > 0 for d in durations)
+
+
+@pytest.mark.parametrize("prefetch_batches", (0, 1))
+def test_dali_like_iterator(monkeypatch, prefetch_batches):
+    class MockDALIGenericIterator(Iterator):
+        def __init__(self, size: int):
+            self.counter = 0
+            self.size = size
+
+        def __next__(self):
+            if self.counter >= self.size:
+                # the DALI iterator needs one extra fetch call to reset the state and stop
+                self.reset()
+                raise StopIteration
+            self.counter += 1
+            return self.counter  # can be anything really
+
+        def reset(self):
+            self.counter = 0
+
+        def __len__(self):
+            return self.size
+
+    import pytorch_lightning.utilities.fetching as fetching
+
+    # the batches without any of our fetching logic
+    expected = [(e, b) for e in range(3) for b in MockDALIGenericIterator(3)]
+
+    nvidia_dali_module = "nvidia.dali.plugin.pytorch"
+    # need different mocking logic based on whether the library is installed
```

Comment:
I didn't want to have to install DALI in the CI for this test, so had to write this voodoo to have the test work both when it's installed and when it isn't